README.markdown
[![Build Status](https://travis-ci.org/mojolingo/es-reindex.svg)](https://travis-ci.org/mojolingo/es-reindex)

# es-reindex - simple ruby gem for copying ElasticSearch index

Simple ruby gem to copy and reindex ElasticSearch index,
copying index settings and mapping(s).

Progress and time estimation is displayed during the scrolling process.

## Requirements

- Ruby 1.9.3 or newer
- Gems:
  - [elasticsearch](https://github.com/elasticsearch/elasticsearch-ruby)

## Usage (command line)

Refer to script's help:

```
$ ./es-reindex.rb -h

Script to copy particular ES index including its (re)creation w/options set
and mapping copied.

Usage:

  ./es-reindex [options] [source_url/]<index> [destination_url/]<index>

    - -r  - remove the index in the new location first
    - -f  - specify frame size to be obtained with one fetch during scrolling
    - -u  - update existing documents (default: only create non-existing)
    - -nm - don't copy mappings and settings
    - optional source/destination urls default to http://127.0.0.1:9200
```

## Usage (in project)

You can also use it as a PORO:

#### To Copy

```ruby
# Options:
# remove: same as -r
# frame: same as -f
# update: same as -u cli option

options = {
  remove: true,
  update: true
}

ESReindex.copy! 'http://my_server/index', 'http://my_server/index_copy', options
```

#### To Reindex

If you want to reindex the destination from the source without copying the mappings/settings from the source, you can do it as such:

```ruby
ESReindex.reindex! 'http://my_server/index', 'http://my_server/index_copy',
  mappings: -> { set_of_mappings },
  settings: -> { set_of_settings}
```

If using the `.reindex!` method, you MUST pass valid mappings/settings in via the options.

#### Callbacks
There also a set of callbacks you can use:

```ruby
ESReindex.copy! 'http://my_server/index', 'http://my_server/index_copy',
  before_create: ->    { do_something },      # Runs before the (re)creation of the destination index
  after_create:  ->    { do_something_else }, # Runs after the (re)creation of the destinatino index
  before_each:   ->doc { use_the doc },       # Runs before each document is copied
  after_each:    ->doc { foo_bar doc },       # Runs after each document is copied
  after_copy:    ->    { finish_thing }       # Runs after everything is copied over
```

#### Callbacks (guards)

You can also use the `:if` or `:unless` callbacks to prevent the copy/reindexing from occuring if conditions are (un)met.  The source client and destination client are passed in:

```ruby
ESReindex.copy! 'http://my_server/index', 'http://my_server/index_copy',
  if:     ->(sclient,dclient) { Time.now.hour > 20 },                 # Only copy the indexes if it's after 8pm
  unless: ->(sclient,dclient) { Time.now.strftime("%A") == "Friday" } # Never copy on Fridays
```

For a more practical example, see the [reindex integration specs](spec/integration/reindex_spec.rb).

## Changelog

The full changelog is available [here](CHANGELOG.markdown).

## Credits

- [Justin Aiken](https://github.com/JustinAiken)
- [Victor Luft](https://github.com/victorluft)

Original script:
  - @geronime
  - @pgaertig

Developed by [Mojo Lingo](http://mojolingo.com).

## License
es-reindex the gem is copyright (c)2014 Mojo Lingo, and released under the terms
of the MIT license. See the LICENSE file for the gory details.

es-reindex is copyright (c)2012 Jiri Nemecek, and released under the terms
of the MIT license. See the LICENSE file for the gory details.


CHANGELOG.markdown
## Changelog

+ __0.3.3__: Added `-nm` option [@wader](https://github.com/wader)
+ __0.3.2__: Remove activesupport dependency entirely, handle aliased indices [@waterlink](https://github.com/waterlink)
+ __0.3.1__: Add activesupport dependency since es-reindex uses methods from it.
+ __0.3.0__: Add `:if` and `:unless` callbacks
+ __0.2.1__: [BUGFIX] Improve callback presence check
+ __0.2.0__: Lots of bugfixes, use elasticsearch client gem, add .reindex! method and callbacks
+ __0.1.0__: First gem release
+ __0.0.9__: Gemification, Oj -> MultiJSON
+ __0.0.8__: Optimization in string concat (@nara)
+ __0.0.7__: Document header arguments `_timestamp` and `_ttl` are copied as well
+ __0.0.6__: Document headers in bulks are now assembled and properly JSON dumped
+ __0.0.5__: Merge fix for trailing slash in urls (@ichinco), formatting cleanup
+ __0.0.4__: Force create only, update is optional (@pgaertig)
+ __0.0.3__: Yajl -> Oj
+ __0.0.2__: repated document count comparison
+ __0.0.1__: first revision


spec/es-reindex_spec.rb
require 'spec_helper'

describe ESReindex do
  let(:src)     { 'http://foo/bar_index' }
  let(:dst)     { 'http://biz/baz_index' }
  let(:options) { {update: true} }

  let(:reindexer) { ESReindex.new src, dst, options }

  it "can be freshly initialized with options" do
    expect(reindexer.options).to eq remove: false, update: true, frame: 1000, from_cli: false, copy_mappings: true, :scroll=>"10m", :query=>{:query=>{:match_all=>{}}}
  end

  context "with invalid callback options" do
    let(:options) { {if: :not_a_proc} }

    it "fails" do
      expect{
        reindexer
      }.to raise_error ArgumentError, "if must be a callable object"
    end
  end

  it "starts with 0 indexes done" do
    expect(reindexer.done).to eq 0
  end

  describe "#copy!" do
    after { reindexer.copy! }

    before do
      %w{
        confirm clear_destination create_destination copy_docs check_docs
      }.each { |meth| reindexer.stub(meth).and_return true }
    end

    context "when run as a PORO" do
      it "doesn't use #exit" do
        expect(reindexer).to receive(:clear_destination).and_return false
        expect(reindexer).not_to receive :exit
      end
    end

    context "when run from the CLI bin" do
      let(:options) { {from_cli: true} }

      it "exits 1 on failure" do
        expect(reindexer).to receive(:clear_destination).and_return false
        expect(reindexer).to receive(:exit).with 1
      end

      it "exits 0 on success" do
        expect(reindexer).to receive(:exit).with 0
      end
    end
  end

  skip "it can actually do stuff"
end


spec/es-reindex/args_parser_spec.rb
require 'spec_helper'

describe ESReindex::ArgsParser do

  let(:parsed_args)  { described_class.parse args }
  let(:parsed_src)   { parsed_args[0] }
  let(:parsed_dst)   { parsed_args[1] }
  let(:parsed_opts)  { parsed_args[2] }

  context "with no src or dst" do
    let(:args) { ["-r", "-u"] }

    it "src is nil" do
      expect(parsed_src).to be_nil
    end

    it "dst is nil" do
      expect(parsed_dst).to be_nil
    end
  end

  context "with a src" do
    let(:args) { ["-r", "-u", "http://foo/index"] }

    it "src is set" do
      expect(parsed_src).to eq 'http://foo/index'
    end

    it "dst is nil" do
      expect(parsed_dst).to be_nil
    end
  end

  context "with a src and dst" do
    let(:args) { ["-r", "-u", "http://foo/index", "bar/index"] }

    it "src is set" do
      expect(parsed_src).to eq 'http://foo/index'
    end

    it "dst is set" do
      expect(parsed_dst).to eq 'bar/index'
    end
  end

  context "without -f" do
    let(:args) { ["-u"] }

    it "sets the frame to the default (1000)" do
      expect(parsed_opts[:frame]).to eq 1000
    end
  end

  context "with -f" do
    let(:args) { ["-f", "1500"] }

    it "sets the frame" do
      expect(parsed_opts[:frame]).to eq 1500
    end
  end

  context "without -u" do
    let(:args) { ["-f", '1000', 'foosrc/index'] }

    it "sets update to false" do
      expect(parsed_opts[:update]).to be false
    end
  end

  context "with -u" do
    let(:args) { ["-u"] }

    it "sets the frame" do
      expect(parsed_opts[:update]).to be true
    end
  end

  context "without -nm" do
    let(:args) { ["-u"] }

    it "sets copy_mappings to true" do
      expect(parsed_opts[:copy_mappings]).to be true
    end
  end

  context "with -nm" do
    let(:args) { ["-u", "-nm"] }

    it "sets copy_mappings to false" do
      expect(parsed_opts[:copy_mappings]).to be false
    end
  end
end


lib/es-reindex.rb
require 'elasticsearch'
require 'logger'

require 'es-reindex/railtie' if defined?(Rails)

class ESReindex

  DEFAULT_URL = 'http://127.0.0.1:9200'

  attr_accessor :src, :dst, :options, :surl, :durl, :sidx, :didx, :sclient, :dclient, :start_time, :done, :settings, :mappings

  def self.copy!(src, dst, options = {})
    self.new(src, dst, options).tap do |reindexer|
      reindexer.setup_index_urls
      reindexer.copy! if reindexer.okay_to_proceed?
    end
  end

  def self.reindex!(src, dst, options={})
    self.new(src, dst, options.merge(copy_mappings: false)).tap do |reindexer|
      reindexer.setup_index_urls
      reindexer.copy! if reindexer.okay_to_proceed?
    end
  end

  def initialize(src, dst, options = {})
    ESReindex.logger ||= Logger.new(STDERR)

    @src     = src || ''
    @dst     = dst || ''
    @options = {
      from_cli: false, # Coming from CLI?
      remove: false, # remove the index in the new location first
      update: false, # update existing documents (default: only create non-existing)
      frame:  1000,  # specify frame size to be obtained with one fetch during scrolling
      copy_mappings: true # Copy old mappings/settings
    }.merge! options

    %w{
      if unless mappings settings before_create after_create before_each after_each after_copy
    }.each do |callback|
      if options.has_key?(callback.to_sym) && !options[callback.to_sym].respond_to?(:call)
        raise ArgumentError, "#{callback} must be a callable object"
      end
    end

    @done = 0
  end

  def setup_index_urls
    @surl, @durl, @sidx, @didx = '', '', '', ''
    [[src, surl, sidx], [dst, durl, didx]].each do |param, url, idx|
      if param =~ %r{^(.*)/(.*?)$}
        url.replace $1
        idx.replace $2
      else
        url.replace DEFAULT_URL
        idx.replace param
      end
    end

    @sclient = Elasticsearch::Client.new host: surl
    @dclient = Elasticsearch::Client.new host: durl
  end

  def okay_to_proceed?
    okay = true
    okay = options[:if].call(sclient, dclient) if options.has_key?(:if)
    okay = (okay && !(options[:unless].call sclient, dclient)) if options.has_key?(:unless)
    log 'Skipping action due to guard callbacks' unless okay
    okay
  end

  def copy!
    log "Copying '#{surl}/#{sidx}' to '#{durl}/#{didx}'#{remove? ? ' with rewriting destination mapping!' : update? ? ' with updating existing documents!' : '.'}"
    confirm if from_cli?

    success = (
      clear_destination  &&
      create_destination &&
      copy_docs          &&
      check_docs
    )

    if from_cli?
      exit (success ? 0 : 1)
    else
      success
    end
  end

  def confirm
    printf "Confirm or hit Ctrl-c to abort...\n"
    $stdin.readline
  end

  def clear_destination
    dclient.indices.delete(index: didx) if remove? && dclient.indices.exists(index: didx)
    true
  rescue => e
    false
  end

  def create_destination
    unless dclient.indices.exists index: didx
      if copy_mappings?
        return false unless get_settings
        return false unless get_mappings
        create_msg = " with settings & mappings from '#{surl}/#{sidx}'"
      else
        @mappings = options[:mappings].nil? ? {} : options[:mappings].call
        @settings = options[:settings].nil? ? {} : options[:settings].call
        create_msg = ""
      end

      options[:before_create] && options[:before_create].call

      log "Creating '#{durl}/#{didx}' index#{create_msg}..."
      dclient.indices.create index: didx, body: { settings: settings, mappings: mappings }
      log "Succesfully created '#{durl}/#{didx}''#{create_msg}."

      options[:after_create] && options[:after_create].call
    end

    true
  end

  def get_settings
    unless settings = sclient.indices.get_settings(index: sidx)
      log "Failed to obtain original index '#{surl}/#{sidx}' settings!"
      return false
    end

    @settings = fetch_index_config(settings, sidx)["settings"]
    @settings["index"]["version"].delete "created"
  end

  def get_mappings
    unless mappings = sclient.indices.get_mapping(index: sidx)
      log "Failed to obtain original index '#{surl}/#{sidx}' mappings!", :error
      return false
    end
    @mappings = fetch_index_config(mappings, sidx)["mappings"]
  end

  def copy_docs
    log "Copying '#{surl}/#{sidx}' to '#{durl}/#{didx}'..."
    @start_time = Time.now

    scroll = sclient.search index: sidx, search_type: "scan", scroll: '10m', size: frame
    scroll_id = scroll['_scroll_id']
    total = scroll['hits']['total']
    log "Copy progress: %u/%u (%.1f%%) done.\r" % [done, total, 0]

    action = update? ? 'index' : 'create'

    while scroll = sclient.scroll(scroll_id: scroll['_scroll_id'], scroll: '10m') and not scroll['hits']['hits'].empty? do
      bulk = []
      scroll['hits']['hits'].each do |doc|
        options[:before_each] && options[:before_each].call
        ### === implement possible modifications to the document
        ### === end modifications to the document
        base = {'_index' => didx, '_id' => doc['_id'], '_type' => doc['_type'], data: doc['_source']}
        bulk << {action => base}
        @done = done + 1
        options[:after_each] && options[:after_each].call
      end
      unless bulk.empty?
        dclient.bulk body: bulk
      end

      eta = total * (Time.now - start_time) / done
      log "Copy progress: #{done}/#{total} (%.1f%%) done in #{tm_len}. E.T.A.: #{start_time + eta}." % (100.0 * done / total)
    end

    log "Copy progress: %u/%u done in %s.\n" % [done, total, tm_len]

    options[:after_copy] && options[:after_copy].call

    true
  end

  def check_docs
    log 'Checking document count... '
    scount, dcount = 1, 0
    begin
      Timeout::timeout(60) do
        while true
          scount = sclient.count(index: sidx)["count"]
          dcount = dclient.count(index: didx)["count"]
          break if scount == dcount
          sleep 1
        end
      end
    rescue Timeout::Error
    end
    log "Document count: #{scount} = #{dcount} (#{scount == dcount ? 'equal' : 'NOT EQUAL'})"

    scount == dcount
  end

  class << self
    attr_accessor :logger
  end

private

  def log(msg, level = :info)
    ESReindex.logger.send level, msg
  end

  def remove?
    @options[:remove]
  end

  def update?
    @options[:update]
  end

  def frame
    @options[:frame]
  end

  def from_cli?
    @options[:from_cli]
  end

  def copy_mappings?
    @options[:copy_mappings]
  end

  def tm_len
    l = Time.now - @start_time
    t = []
    t.push l/86400; l %= 86400
    t.push l/3600;  l %= 3600
    t.push l/60;    l %= 60
    t.push l
    out = sprintf '%u', t.shift
    out = out == '0' ? '' : out + ' days, '
    out << sprintf('%u:%02u:%02u', *t)
    out
  end

  # Accounts for aliased indices. When index is aliased there will not
  # be a key nor in settings, nor in mappings. But there will be the
  # only key for original index name.
  def fetch_index_config(config, index)
    config.fetch(index) { config.values.first }
  end

end
